### PR TITLE
Update JSON-RPC

### LINKED ISSUE #10264: Update JSON-RPC and Related Dependencies
### Feature Description:

We use `vscode-ws-jsonrpc` a [wrapper around `vscode-jsonrpc`](https://github.com/TypeFox/vscode-ws-jsonrpc) developed by TypeFox that TypeFox has stopped supporting. That package currently only supports `vscode-jsonrpc` < 6 as `vscode-jsonrpc@6` introduced significant changes, and that affects the versions of the language server protocol that we can use. A [fork](https://github.com/CodinGame/monaco-jsonrpc) has been made of the `vscode-ws-jsonrpc` that has updated to use `vscode-jsonrpc` >= 6, and we should consider updating to use that package instead or dump the wrapper and use `vscode-jsonrpc` directly.

### PR DESCRIPTION
**Pull request by colin-grant-work**
#### What it does

Fixes #10264 by switching to use a fork of `vscode-ws-jsonrpc` that has been updated to support `vscode-jsonrpc@6`. 

I could use some help, though, as I'm not very familiar with all of the details of the JSON RPC system. I'll make some comments pointing to the places where I'm uncertain of the right move.

#### How to test

1. Since this modifies the core of our communication with the backend, thoroughly test... everything, I guess.
2. In particular, check that interactions with plugins work as expected.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by tsmaeder** [comment]
T.b.h, I would tend toward in-sourcing this code: I believe the newest version of vscode-jsonrpc actually _has_ support for sockets/ipc, etc. We would be better served maintaining the bit of glue code needed ourselves.

**Comment by paul-marechal** [comment]
@msujew good catch, this is now fixed. There is still an error being thrown when stopping the session but it already happens on master so that makes it out of scope for this PR.

**Comment by colin-grant-work** [comment]
@tsmaeder, it seems like there's some overlap between the work here and the work you've been doing on the debug system. Would you mind taking a look at this so we can merge if appropriate, and you can then use some of the new messaging equipment in your PR?

File: packages/plugin-ext/src/common/connection.ts
Comment (tsmaeder) [inline packages/plugin-ext/src/common/connection.ts:96]: I'm not sure we need a deprecation cycle here: it's not "official" API and I doubt someone actually used this implementation detai.

File: packages/plugin-ext/src/common/connection.ts
Comment (tsmaeder) [inline packages/plugin-ext/src/common/connection.ts:60]: We're wrapping a string channel around a json rcp connection which wraps a string channel. That does not really make sense, IMO. I got rid of that all in https://github.com/eclipse-theia/theia/pull/10333. So I wouldn't sweat the small stuff here.

File: packages/plugin-ext/src/common/plugin-message-writer.ts
Comment (tsmaeder) [reply packages/plugin-ext/src/common/plugin-message-writer.ts:37]: @colin-grant-work like what?

File: packages/core/src/node/messaging/messaging-contribution.ts
Comment (tsmaeder) [inline packages/core/src/node/messaging/messaging-contribution.ts:82]: Is this even ever used? I didn't find any usages.

**Review by tsmaeder — DISMISSED** [review]
I really wish we could have done the changing of the imports to 'vscode-languageserver-protocol'  in a separate PR: 88 files is very big when "real changes" are insterspersed with just changed imports.

The only real change I'm requesting is the renaming of `TheiaMessagingConnection`: I just find the name very confusing. There are many things in this area of code that I disapprove of, but they were there before so 🤷 .

In the long run, I would like to get rid of jsonrpc anyway: we already have a remote proxy invocation framework in our code and I don't think we need two: the only thing I can see missing is the ability to read partial messages, but that could quite simply be retrofitted.

File: packages/plugin-ext/src/common/plugin-message-writer.ts
Comment (colin-grant-work) [reply packages/plugin-ext/src/common/plugin-message-writer.ts:37]: If you don't know, and @paul-marechal doesn't know, then I don't know. The method is required for the interface, but it doesn't seem we have anything we want to do in this case.

**Comment by colin-grant-work** [comment]
> @colin-grant-work do we need more testing on this PR?

@tsmaeder, more testing never hurts. It seems that @msujew was happy with his last run through, and I haven't noticed any problems, but you may be aware of use cases that we're not hitting.

File: packages/core/package.json
Comment (paul-marechal) [reply packages/core/package.json:65]: It's a struggle.

File: packages/plugin-ext/src/common/connection.ts
Comment (paul-marechal) [reply packages/plugin-ext/src/common/connection.ts:60]: I did what I must to get it to compile as before, I'd be more than happy to merging #10333 after this!

File: packages/core/src/node/messaging/messaging-contribution.ts
Comment (paul-marechal) [reply packages/core/src/node/messaging/messaging-contribution.ts:82]: Not anymore but despite the amount of changes in this PR I tried to only touch what was broken/bad.

This method is not bothering anyone so I'd keep it for now :)

File: packages/plugin-ext/src/common/connection.ts
Comment (paul-marechal) [reply packages/plugin-ext/src/common/connection.ts:96]: I'm very fine with that, will remove.

File: packages/core/src/common/messaging/proxy-factory.ts
Comment (paul-marechal) [reply packages/core/src/common/messaging/proxy-factory.ts:275]: Like I mentioned, `undefined` values are transformed into `null` without the type system knowing.

https://github.com/eclipse-theia/theia/blob/8bf15dc1d9b186b11c1f954b4c51500e611d3072/packages/plugin-ext/src/main/common/plugin-paths-protocol.ts#L24-L25

Here `workspaceUri` is actually `string | null` when used over RPC. We got "lucky" here as the only implementation checks for truthiness to detect the type rather than some `workspaceUri === undefined` which would not work.

One solution would be for interfaces meant to be used over RPC to almost never use `undefined` in their API.

File: packages/core/src/common/messaging/theia-message-connection.ts
Comment (colin-grant-work) [reply packages/core/src/common/messaging/theia-message-connection.ts:26]: Since its job is to forward messages, how about `RPCMessageRelay`?

**Comment by tsmaeder** [comment]
Btw: does the new version help with https://github.com/eclipse-theia/theia/issues/9514?

**Comment by colin-grant-work** [comment]
> Btw: does the new version help with #9514?

@tsmaeder, I haven't checked, but it would surprise me if it had much effect on the performance - I don't think there was that much extraneous code running before, and I don't think there's much less now. We still have to JSON-ify everything and send it over the wire (short or long) to the other end. There's a little bit of argument processing on our side and in the package, and that certainly adds a little overhead, but I don't think that's changed much. Someone please bring in a performance tracer and correct me :-).

File: packages/core/src/common/messaging/theia-message-connection.ts
Comment (tsmaeder) [inline packages/core/src/common/messaging/theia-message-connection.ts:26]: I guess the file name does not make sense now.

**Comment by msujew** [comment]
@colin-grant-work @paul-marechal I assume a rebase is in order due to #10333. I would retest this afterwards and approve if possible. We'd have a few weeks after merging to find possible regressions before the next release.

**Comment by colin-grant-work** [comment]
I'm going to close this for now, since it looks like there's more interest in rewriting our message layer than in updating this package.

**Pull request closed without merging**
